internal/n/reliable_notifier.go

```diff
@@ -0,0 +1,205 @@
+package n
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/capatazlib/go-capataz/internal/s"
+)
+
+// rootName is the name of the ReliableNotifier supervisor root tree node
+var rootName = "reliable-notifier"
+
+// notifierSettings contains settings and callbacks for a ReliableNotifier
+// instance
+type notifierSettings struct {
+	entrypointBufferSize    uint
+	notifierTimeoutDuration time.Duration
+
+	onReliableNotifierFailure func(error)
+	onNotifierTimeout         func(string)
+}
+
+// ReliableNotifierOpt allows clients to tweak the behavior of a
+// ReliableNotifier instance
+type ReliableNotifierOpt func(*notifierSettings)
+
+// newNotifierWorker runs a worker that listens to a channel dedicated to the
+// given event notifier. In the situation the notifierFn panics, the worker gets
+// restarted.
+func newNotifierWorker(
+	_ notifierSettings,
+	name string,
+	notifierFn s.EventNotifier,
+) (chan s.Event, s.Node) {
+	// ch := make(chan s.Event, settings.notifierBufferSize)
+	ch := make(chan s.Event)
+	return ch, s.NewWorker(
+		name,
+		func(ctx context.Context) error {
+			for {
+				select {
+				case <-ctx.Done():
+					return nil
+				case ev := <-ch:
+					notifierFn(ev)
+				}
+			}
+		},
+	)
+}
+
+// newNotifierSubTree runs a subtree of notifier workers. It restarts all
+// notifiers if one of them fails above restart threshold.
+func newNotifierSubTree(
+	settings notifierSettings,
+	notifierFns map[string]s.EventNotifier,
+) (s.Node, map[string](chan s.Event)) {
+
+	workers := make([]s.Node, 0, len(notifierFns))
+	notifierChans := make(map[string](chan s.Event))
+
+	for name, notifierFn := range notifierFns {
+		ch, worker := newNotifierWorker(settings, name, notifierFn)
+		notifierChans[name] = ch
+		workers = append(workers, worker)
+	}
+	notifierTree := s.Subtree(s.NewSupervisorSpec("notifiers", s.WithNodes(workers...)))
+
+	return notifierTree, notifierChans
+}
+
+// runEntrypointListener listens to a channel for events, and it then broadcast
+// in a non-blockin fashion the same event across multiple workers.
+func runEntrypointListener(
+	ctx context.Context,
+	settings notifierSettings,
+	entrypointCh chan s.Event,
+	notifierChans map[string](chan s.Event),
+) error {
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+
+		case ev := <-entrypointCh:
+			for name, ch := range notifierChans {
+				notifyCtx, stopTimer := context.WithTimeout(
+					context.Background(),
```

Comment:
I did that first, but I wanted to rely on the returned cleanup function only to stop this system. If we use the same context that the Supervision Tree would use, then it would happen that this notification system tree gets shutdown before the application tree gets shutdown, does that make sense?